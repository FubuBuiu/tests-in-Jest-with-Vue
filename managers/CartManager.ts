import Vue from 'vue';
import { GlobalTypes } from '@/types/global-types';

export default {
  install: (Vue: any) => {
    // istanbul ignore next
    Vue.prototype.$cart = new CartManager();
  },
};

export interface CartProduct extends GlobalTypes.Product {
  quantity: number;
}

const initialState = {
  open: false as boolean,
  productList: [] as CartProduct[],
};

export class CartManager {
  state;

  constructor() {
    this.state = Vue.observable(initialState);
  }

  open() {
    this.state.open = true;
  }

  close() {
    this.state.open = false;
  }

  getState() {
    return this.state;
  }

  existProductInTheCart(product: GlobalTypes.Product): boolean {
    return !!this.state.productList.find(({ id }) => product.id === id);
  }

  increaseQuantity(index: number) {
    this.state.productList[index].quantity++;
  }

  decreaseQuantity(index: number) {
    this.state.productList[index].quantity--;
  }

  addProduct(product: GlobalTypes.Product) {
    if (this.existProductInTheCart(product)) {
      this.state.productList.find(({ id }) => product.id === id)!.quantity++;
      return;
    }
    const productAux = {
      ...product,
      quantity: 1,
    };
    this.state.productList.push(productAux);
  }

  removeProduct(productId: string) {
    this.state.productList = [
      ...this.state.productList.filter((product) => product.id !== productId),
    ];
  }

  hasProduct(): boolean {
    return this.state.productList.length > 0;
  }

  clearCart() {
    this.state.productList = [];
  }
}
