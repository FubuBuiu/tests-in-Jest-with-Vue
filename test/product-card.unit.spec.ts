import { expect } from '@jest/globals';
import { mount, createLocalVue } from '@vue/test-utils';
import Vue from 'vue';
import Vuetify from 'vuetify';
import ProductCard from '@/components/product-card.vue';
import { makeServer } from '~/miragejs/server';
import { CartManager } from '~/managers/CartManager';

Vue.use(Vuetify);

let localVue: any;
let vuetify: any;
let server: any;

function mountProductCard() {
  const product = server.create('product', {
    image:
      'https://images.unsplash.com/photo-1587466412525-87497b34fc88?ixlib=rb-4.0.3&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=1173&q=80',
    title: 'Relógio bonito',
    price: '45.00',
  });

  const cartManager = new CartManager();

  const wrapper = mount(ProductCard, {
    localVue,
    vuetify,
    stubs: { Nuxt: true },
    mocks: {
      $cart: cartManager,
    },
    propsData: {
      product,
    },
  });
  return {
    wrapper,
    product,
    cartManager,
  };
}
describe('ProductCard component - unit', () => {
  beforeEach(() => {
    localVue = createLocalVue();
    vuetify = new Vuetify();
    server = makeServer({ environment: 'test' });
  });

  afterEach(() => {
    server.shutdown();
  });

  test('should match snapshot', () => {
    const { wrapper } = mountProductCard();

    expect(wrapper.element).toMatchSnapshot();
  });
  test('should mount the component', () => {
    const { wrapper } = mountProductCard();

    expect(wrapper.vm).toBeDefined();
    expect(wrapper.text()).toContain('Relógio bonito');
    expect(wrapper.text()).toContain('45.00');
  });
  test('should add product to cart when button gets clicked', async () => {
    const { wrapper, cartManager, product } = mountProductCard();

    const spy1 = jest.spyOn(cartManager, 'open');
    const spy2 = jest.spyOn(cartManager, 'addProduct');
    await wrapper.find('button').trigger('click');

    expect(spy1).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledTimes(1);
    expect(spy2).toHaveBeenCalledWith(product);
  });
});
