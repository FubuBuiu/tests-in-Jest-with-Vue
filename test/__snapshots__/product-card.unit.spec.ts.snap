// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component - unit should match snapshot 1`] = `
<div
  class="v-card v-sheet theme--light"
  style="min-width: 300px; max-width: 300px;"
>
  <div
    class="v-image v-responsive theme--light"
    style="height: 190px;"
  >
    <transition-stub
      mode="in-out"
      name="fade-transition"
    >
      <div
        class="v-image__image v-image__image--preload v-image__image--cover"
        style="background-position: center center;"
      />
    </transition-stub>
    <div
      class="v-responsive__content"
    />
  </div>
   
  <div
    class="v-card__title"
  >
    <div
      class="row"
    >
      <div
        class="text-truncate col"
      >
        <span
          class="v-tooltip v-tooltip--bottom"
        >
          <!---->
          <span
            aria-expanded="false"
            aria-haspopup="true"
          >
            
              Rel√≥gio bonito
            
          </span>
        </span>
      </div>
       
      <div
        align="right"
        class="pa-0 col col-3"
      >
        <button
          class="v-btn v-btn--is-elevated v-btn--fab v-btn--has-bg v-btn--round theme--light v-size--small blue"
          data-testid="addProductToCartButton"
          dense=""
          style="transform: translateY(-50%);"
          type="button"
        >
          <span
            class="v-btn__content"
          >
            <i
              aria-hidden="true"
              class="v-icon notranslate pa-0 ma-0 mdi mdi-cart-outline theme--light white--text"
            />
          </span>
        </button>
      </div>
    </div>
  </div>
   
  <div
    class="v-card__subtitle"
  >
     $ 45.00 
  </div>
</div>
`;
